From 69fa4e94da5f0f88d9ea4de1907f8fbc20367cf2 Mon Sep 17 00:00:00 2001
From: Andrei Stefanescu <andrei.stefanescu@nxp.com>
Date: Mon, 29 Aug 2022 15:01:10 +0300
Subject: [PATCH 099/102] net: pfeng: handle memory leaks

The `pfeng_hw_init_ring` allocates various resources which were not
de-allocated in case of error.

Issue: ALB-9227

Upstream-Status: Pending 

Signed-off-by: Andrei Stefanescu <andrei.stefanescu@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/net/pfeng/pfeng_hw.c | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/drivers/net/pfeng/pfeng_hw.c b/drivers/net/pfeng/pfeng_hw.c
index be6b98cada..cbb4abecce 100644
--- a/drivers/net/pfeng/pfeng_hw.c
+++ b/drivers/net/pfeng/pfeng_hw.c
@@ -746,7 +746,7 @@ static struct pfe_hif_ring *pfeng_hw_init_ring(bool is_rx)
 
 	if (!ring->bd) {
 		pr_warn("HIF ring couldn't be allocated.\n");
-		return NULL;
+		goto err_with_ring;
 	}
 
 	mmu_set_region_dcache_behaviour((phys_addr_t)ring->bd,
@@ -757,7 +757,7 @@ static struct pfe_hif_ring *pfeng_hw_init_ring(bool is_rx)
 
 	if (!ring->wb_bd) {
 		pr_warn("HIF ring couldn't be allocated.\n");
-		return NULL;
+		goto err_with_bd;
 	}
 
 	mmu_set_region_dcache_behaviour((phys_addr_t)ring->wb_bd,
@@ -775,10 +775,8 @@ static struct pfe_hif_ring *pfeng_hw_init_ring(bool is_rx)
 		ring->mem = memalign(page_size,
 				     PFE_BUF_SIZE * PFE_HIF_RING_CFG_LENGTH);
 		offset = ring->mem;
-		if (!ring) {
-			free(ring);
-			return NULL;
-		}
+		if (!ring->mem)
+			goto err_with_wb_bd;
 	}
 
 	for (ii = 0; ii < RING_LEN; ii++) {
@@ -816,6 +814,14 @@ static struct pfe_hif_ring *pfeng_hw_init_ring(bool is_rx)
 	      ring->bd, ring->wb_bd, ring->mem);
 
 	return ring;
+
+err_with_wb_bd:
+	free(ring->wb_bd);
+err_with_bd:
+	free(ring->bd);
+err_with_ring:
+	free(ring);
+	return NULL;
 }
 
 struct pfe_hw_chnl *pfeng_hw_init_chnl(void)
-- 
2.17.1

